### ISSUE TITLE
persisting data

### ISSUE THREAD

**Issue by jeremybmerrill** [a-mma/AquilaDB#31]
Hi @freakeinstein, it's me again! :)

How do I get AquilaDB to persist the FAISS db to disk? I'm trying to be able to restart the underlying AWS instance (and thus the AquilaDB Docker container) and have my data persist.

It doesn't seem like the data does persist. Even after I set up a working example, no file ever shows up in `/data/VDB`. Is this a bug? Am I missing something?

I'm running off of the `master` branch (having implemented the b64-related change myself).

Thanks!

**Comment by freakeinstein** [comment]
Hi @jeremybmerrill 
Let me break down everything in detail.

First of all, Annoy is there to support AquilaDB, when FAISS is not available.
* Based on current configuration, FAISS requires at least around 10000 vectors to be indexed first to start using it for k-NN retrieval.
* Of Course, FAISS is better than Annoy and is best used for bigger datasets. But for smaller datasets (some AquilaDB users might have less data to be indexed, if we rely on FAISS only AquilaDB is no more useful to them), Annoy automatically kicks in.
* As soon as the dataset size exceeds 10000 examples, FAISS will automatically kick in replacing annoy. So no one need to worry about that.
* You can configure any of these with DB_config.yml file (before starting AquilaDB).


Data persistence
* If you check Annoy source code, it was configured to persist data to the disk. The code was commented out recently, because of the frequent disk write it required (so, DB write is slow). I have re-enabled it again (check above pull request) and decided to explain best practices to the users for better performance.
* FAISS will also keep data on disk when it is activated.

Best practices
* Because of the current implementation of `vector indexing` that AquilaDB have (one by one vectors are written to the index), a couple of values should be configured in DB_config.yml file before starting AquilaDB.
* Approximately how many documents are you going insert into DB in bulk (within a short amount of time)? Let's say, you have 'n' documents to be inserted to AquilaDB ASAP before start testing it and after that, you might send only a couple of them in regular intervals. (This is a common pattern when we build a product. We might have a dataset at our reserve and we need to load them all to the database on the first run. After that, the database will be filled with real user data and that will be regular and relaxed insertion). So, set the value of `faiss.init.vecount` inside `DB_config.yml file` to around `n`. if n is higher than 10000, then it is good to keep the value 10000 for the sake of your RAM.
* Also if you can guess the maximum length (dimension) of your vector, just do this calculation: `d = dimension + (dimension % 8)`. And set the value of `faiss.init.vd` inside `DB_config.yml file` to exact `d`.

**Comment by freakeinstein** [comment]
Fixed. Merged into develop. Closing.

**Comment by freakeinstein** [comment]
> So I should set `vecount` and `vd` and that'll be enough to get it to persist the data?
vecount - is the minimum number of documents to be inserted to start first indexing in either Annoy of FAISS (chosen based on the 10000 boundary)
vd - is not important in this scenario, but a useful parameter to be considered because your input vectors will be resized to this size before indexing.

> What happens when I add additional documents beyond the FAISS_MIN_LIMIT
Then each vector insertion will be recorded into the disk. It's not an efficient method. Wanted to make it better in the near future.

> I wonder if it might be better to have an option for the client 
It's a good suggestion. Currently, trying to make everything less complicated by abstracting much details from the user. That's why the values were moved into configs and triggers were automated based on threshold values. Looking forward to incorporating this suggestion in an easy way.

**Comment by freakeinstein** [comment]
>Looks like the default Annoy vector size needs to be set somewhere too
Could you please give me some more details on this? What were the values you set?

**Comment by freakeinstein** [comment]
That wouldn't be a problem here because, it is getting replaced at https://github.com/a-mma/AquilaDB/blob/19355500be59e539ea3e210b8de3ae169470fba5/src/hannoy/index.py#L16 by the actual value we have set in the DB_config.yml during Annoy Initialisation. Something else might be causing that. Let me check it.

**Comment by freakeinstein** [comment]
Thanks for your support, I appreciate it. I was able to reproduce the issue. It seems weird. Let me spend some time on it.

**Comment by freakeinstein** [comment]
I have added a Question / Issue in Annoy's repo. 

In our code we used `unload()` method that Annoy have - which is not working the intended way (that was our mistake to use that method). When we call `unload()` it actually clears the old data from the index. That's why we were seeing misbehaviours in the kNN response. 

It seems like Annoy doesn't have an option to add new data to loaded index. I have asked for their support here: https://github.com/spotify/annoy/issues/411
